organic_scoring/organic_scoring_base.py

```diff
@@ -0,0 +1,319 @@
+        rewards: dict[str, Any],
+        sample: Any,
+        *args,
+        **kwargs,
+    ) -> dict[str, Any]:
+        """Log the results of the organic scoring iteration
+
+        Args:
+            logs: The logs to record. Default values in the dict:
+                - "time_sample": Time taken in seconds to sample the organic queue or synthetic dataset;
+                - "time_responses": Time taken in seconds to concurrently query the miners and generate reference;
+                - "time_rewards": Time taken in seconds to generate rewards;
+                - "time_weights": Time taken in seconds to set the weights;
+                - "time_total": Total time taken in seconds for the iteration;
+                - "organic_queue_len": Current length of the organic queue;
+                - "is_organic_sample": If the sample is from the organic queue.
+            reference: The reference data.
+            responses: The responses from the miners.
+            rewards: The generated rewards.
+            sample: The sample used.
+
+        Returns:
+            dict[str, Any]: The logs recorded.
+        """
+        return logs
+
+    def _start_run_loop(self):
+        """Start the run loop for the organic scoring task"""
+        self._axon.attach(
+            forward_fn=self._on_organic_entry,
+            blacklist_fn=self._blacklist_fn,
+            priority_fn=self._priority_fn,
+            verify_fn=self._verify_fn,
+        )
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+        try:
+            loop.run_until_complete(self._run_loop())
+        finally:
+            loop.close()
+
+    async def _run_loop(self):
+        """The main loop for running the organic scoring task, either based on a time interval or steps"""
+        while not self._should_exit:
+            if self._trigger == "steps":
+                while self._step_counter < self._trigger_frequency:
+                    await asyncio.sleep(0.1)
+
+            timer_total = time.perf_counter()
+
+            timer_sample = time.perf_counter()
+            is_organic_sample = False
+            if not self._organic_queue.is_empty():
+                # Choose random organic sample.
+                sample = self._organic_queue.sample()
+                is_organic_sample = True
+            else:
+                # Choose if organic queue is empty, choose random sample from provided datasets.
+                sample = random.choice(self._synth_dataset).sample()
+
+            timer_sample_elapsed = time.perf_counter() - timer_sample
+
+            # Concurrently generate reference and query miners.
+            timer_responses = time.perf_counter()
+            reference_task = asyncio.create_task(self._generate_reference(sample))
+            responses_task = asyncio.create_task(self._query_miners(sample))
+            reference, responses = await asyncio.gather(reference_task, responses_task)
```

Comment:
why is it important here that they run concurrently? I'm sure you have a good reason, but I don't fully understand why it's needed:)

Reply:
Since reference is generated with local LLM, while miners are queries over the network, these 2 steps are independent on each other so it can and should be done concurrently

Reply:
I get that, but then why use create_task rather than simply
```
            reference_task = self._generate_reference(sample)
            responses_task = self._query_miners(sample)
            reference, responses = await asyncio.gather(reference_task, responses_task)
```

If you just want them to run independently at the same-ish time but don't care about exact concurrance?

I'm not an expert on asyncio so I'm probably missing something!:)

Reply:
That's a good point, it would look more simple. The only difference is that `create_task` explicitly schedules the tasks to run concurrently as soon as they are created, even before `await asyncio.gather`. Your approach will run them only once `gather` is called. In practice the difference should be unnoticeable tho

Reply:
Got it, that makes sense. Yeah I wasn't sure whether I wasn't understanding something fundamentally!